DOCSP-37507 Databases and Collections (#63)
mongodb/docs-pymongo: source/databases-collections.txt

- **Documents**: Contain literal data such as string, numbers, dates, and other
  embedded documents.

For more information about document field types and structure, see the
:manual:`Documents </core/document/>` guide in the MongoDB Server manual.

.. TODO: Add a diagram here

Access a Database
-----------------

Access a database by using dictionary-style access on your ``MongoClient`` instance.

The following example accesses a database named "test_database":

.. code-block:: python

   database = client["test_database"]

Access a Collection
-------------------

Access a collection by using dictionary-style access on an instance of your database.

The following example accesses a collection named "test_collection":

.. code-block:: python
   :emphasize-lines: 2

   database = client["test_database"]
   collection = database["test_collection"]

.. tip::

   If the provided collection name does not already exist in the database,
   MongoDB implicitly creates the collection when you first insert data
   into it.

Create a Collection
-------------------

Use the ``create_collection()`` method to explicitly create a collection in a
MongoDB database.

> The following example creates a collection called "example_collection":

Reviewer: S:

```suggestion
The following example creates a collection called ``"example_collection"``:
```

   documents and all indexes within that collection.

   Drop a collection only if the data in it is no longer needed.

Configure Read and Write Operations
-----------------------------------

You can control how the driver routes read operations by setting a **read preference**.
You can also control options for how the driver waits for acknowledgment of
read and write operations on a replica set by setting a **read concern** and a
**write concern**.

By default, databases inherit these settings from the ``MongoClient`` instance,
and collections inherit them from the database. However, you can change these
settings on your database or collection by using one of the following methods:

- ``get_database()``: Gets the database and applies the client's read
   preference, read concern, and write preference.
- ``database.with_options()``: Gets the database and applies its current read
   preference, read concern, and write preference.
- ``get_collection()``: Gets the collection and applies its current read preference, read concern, and write preference.
- ``collection.with_options()``: Gets the collection and applies the database's read

>    preference, read concern, and write preference.

Reviewer: These are rendering with extra bold text. Is there an extra space on the second line of each item?

   :name: genre
   :values: reference

.. meta::
   :keywords: table, row, organize, storage

Overview
--------

In this guide, you can learn how to use MongoDB databases and
collections with {+driver-short+}.

MongoDB organizes data into a hierarchy of the following levels:

- Databases
- Collections
- Documents

Databases are the top level of data organization in a MongoDB instance.
MongoDB organizes databases into collections, which contain **documents**.
Documents contain literal data such as strings, numbers, dates, and
other embedded documents. For more information about document field
types and structure, see the :manual:`Documents </core/document/>` guide in the
MongoDB Server manual.

Access a Database
-----------------

Access a database by using dictionary-style access on your ``MongoClient`` instance.

The following example accesses a database named "test_database":

.. code-block:: python

   database = client["test_database"]

Access a Collection
-------------------

Access a collection by using dictionary-style access on an instance of your database.

The following example accesses a collection named "test_collection":

.. code-block:: python
   :emphasize-lines: 2

>    database = client.getDatabase["test_database"]

Author: Good call, that was a bad edit on my part

collections in the database and their associated metadata.

The following example calls the ``list_collections()`` method and iterates over
the cursor to print the results:

.. code-block:: python

   collection_list = database.list_collections()

   for c in collection_list:
       print(c)

To query for only the names of the collections in the database, call the
``list_collection_name()`` method as follows:

.. code-block:: python

   collection_list = database.list_collection_names()

   for c in collection_list:
       print(c)

For more information about iterating over a cursor, see :ref:`pymongo-cursors`. 

Delete a Collection
-------------------

You can delete a collection from the database using the ``drop_collection()``
method.

The following example deletes the ``test_collection`` collection:

.. code-block:: python

   collection = database["test_collection"];
   collection.drop();

.. warning:: Dropping a Collection Deletes All Data in the Collection

   Dropping a collection from your database permanently deletes all
   documents and all indexes within that collection.

   Only drop collections that contain data that is no longer needed.

> Configure Read and Write Operations

Author: I was originally going to split section up, but it started to seem like a lot of repetition so I went back to having them combined (which I also stole from Node). If you think it would be more clear and worth some repetition though I can split it

> Configure Read and Write Operations

Reviewer: Probably okay as is. I didn't realize there was only one database operation

-----------------------------------

You can control how the driver routes read operations by setting a **read preference**.
You can also control options for how the driver waits for acknowledgment of
read and write operations on a replica set by setting a **read concern** and a
**write concern**.

By default, databases inherit these settings from the ``MongoClient`` instance,
and collections inherit them from the database. However, you can change these

> settings on your database or collection by using one of the following methods:

Author: Dictionary-style access uses the default (which is actually when attribute-style does too), but I don't think you can change any of the settings with attribute-style which is why I didn't mention it here

collection called ``test-collection`` with the ``get_collection()`` method:

.. code-block:: python

   database.get_collection("test-collection", 
                       read_preference=ReadPreference.SECONDARY,
                       read_concern="local",
                       write_concern="majority")

The following example shows how to change read and write settings of a
collection called ``test-collection`` with the ``with_options()`` method:

.. code-block:: python

   collection.with_options(read_preference=ReadPreference.SECONDARY,
                           read_concern="local",
                           write_concern="majority")

.. tip::

   To see the types of read preferences available in the ``ReadPreference`` enum, see the
   `API documentation <{+api-root+}pymongo/read_preferences.html#pymongo.read_preferences.ReadPreference>`__.

To learn more about the read and write settings, see the following guides in the
MongoDB Server manual:

- :manual:`Read Preference </core/read-preference/>`
- :manual:`Read Concern </reference/read-concern/>`
- :manual:`Write Concern </reference/write-concern/>`

Tag Sets
~~~~~~~~

> You can :manual:`tag </core/read-preference-tags/>` replica-set

Reviewer: Specify this is MDB functionality

```suggestion
In MongoDB Server, you can :manual:`tag </core/read-preference-tags/>` replica-set
```

> You can :manual:`tag </core/read-preference-tags/>` replica-set
> members according to any criteria you choose. By default, {+driver-short+} ignores tags

Reviewer: slightly expanded explanation

```suggestion
You can apply key-value :manual:`tags </core/read-preference-tags/>` to replica-set
members according to any criteria you choose. By default, {+driver-short+} ignores tags
```

> You can :manual:`tag </core/read-preference-tags/>` replica-set
> members according to any criteria you choose. By default, {+driver-short+} ignores tags

Reviewer: explain purpose

```suggestion
You can :manual:`tag </core/read-preference-tags/>` replica-set
members according to any criteria you choose. You can then use
those tags to target one or more members for a read operation. 

By default, {+driver-short+} ignores tags
```

> when choosing a member to read from, but you can configure your read preference with
> a ``tag_sets`` parameter.

Reviewer: This was a little confusing. Maybe break it up:

```suggestion
when choosing a member to read from. To instruct {+driver-short+}
to prefer certain tags, pass the ``tag_sets`` parameter to the
``_ServerMode`` constructor.
```

[_ServerMode](https://github.com/mongodb/mongo-python-driver/blob/167b9648caf94d57ee618ed240d7910f4553b55e/pymongo/read_preferences.py#L105)

Author: Changed the wording here slightly since I think it was a little bit off originally

The ``tag_sets`` parameter is a list of dictionaries. Each dictionary
includes tag values that the replica-set member must match to be selected for use.
{+driver-short+} attempts each set of tags, in order, until it finds a set of

> tags with at least one matching replica-set member.

Reviewer: S: Cut this explanation. I was thinking about how to clarify it, but the server docs (which we already linked to) already do that. the previous paragraph mentions briefly what tags are, then the next one shows how to specify them with pymongo. i think that's sufficient.

> In the following code example, the tag set passed to the ``read_preference`` parameter

Reviewer: ```suggestion
In the following code example, the tag set passed as the ``read_preference`` parameter
```

instructs {+driver-short+} to prefer reads from the

> New York data center (``dc``) and to fall back to the San Francisco data center:

Reviewer: ```suggestion
New York data center (``'dc': 'ny'``) and to fall back to the San Francisco data center (``'dc': 'sf'``):
```

.. code-block:: python

   db = client.get_database(
        'test', read_preference=Secondary([{'dc': 'ny'}, {'dc': 'sf'}]))

When performing a read, the ``MongoClient`` in the preceding example attempts to find
secondary members in New York, then in San Francisco. If it can't find a secondary member

> in either location, it raises an ``AutoReconnect`` error.

Reviewer: S: Cut this

To avoid an ``AutoReconnect`` error, include an empty dictionary (``{}``) at the end of
the list. This instructs {+driver-short+} to read from any member that matches the mode,

> ignoring tags.

Reviewer: S: Move to Troubleshooting section
